.githooks/pre-commit/format.sh

```diff
@@ -0,0 +1,20 @@
+
+#!/bin/bash -e
+
+#finding clang-format
+CLANG_FORMAT=${CLANG_FORMAT:-$(which clang-format)}
+
+#asserting clang format binnary
+if [ ! -x $CLANG_FORMAT ]; then
+    echo "Failed to format clang-format wasn't found."
+    exit -1
+fi
+
+#finding root directory
+ROOT_DIRECTORY=${ROOT_DIRECTORY:-$(realpath $(pwd))}
+
+# formatting all code
+clang-format -i -style=Mozilla $(find $ROOT_DIRECTORY -type d -name third-party* -prune -o -print | grep -E '.*\.(h|hpp|cc|cpp|cxx)$')
```

Comment:
https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style

Style looks ok, but will clang-format force the renaming of existing tokens that violate this style guide?

IMHO, we should keep this PR at the ready, but methodically reformat the code, a few files at a time (like a mated pair of .h/.cc files).  Once everything is close to the official format, then we commit this PR and make it official/formal/enforced.

It would be nice if there were a CMake create makefile target that reformats code for us.  "make clang-tidy" or something, that we can run on our local workspaces after editing code and before creating a pull request (so we can see the canonical diff first).  Maybe git has a comment to re presubmit hooks but not create a PR that I am unaware of?

One last thing: I don't want to stomp on any pending PRs that Kip, Tim or Igor (or anyone else) is working on.  But I have the contact info for all 4 of you, so I can email you all to get signoff before submitting this PR.

I am excited for this.  The code is a mess and a uniform style will improve readability.  I just want to proceed with caution.

Reply:
1 - Now it is using overcommit to check, not to format.
2 - I created a new target format that can be called as "make format".
3 - I modified the name of test_memcheck to memcheck
4 - I created a new cmake option called PISTACHE_CREATE_OSS_CONTRIB_TARGETS
 - Turns Valgrind, cppcheck and clang-format as required dependencies and configure them
 - Creates format target
 - Create test and docs targets
 - this option is enabled by default when using debug target

What do you think about these modifications?